{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exam 2 answers - Guillermo Chumaceiro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To solve this problem I'm going to use the csv library to read both files and manipulate them in code.\n",
    "\n",
    "To abstract the problem as much as possible, I created a function called control_sequence that makes sure that the first column of two files (can be .csv or .tsv) have the same sequence.\n",
    "\n",
    "This function opens both files and then reads them with the csv library using the delimiter specified when the function is called.\n",
    "\n",
    "I used csv.reader to generate lists (instead of dictionaries with csv.DictReader) because that way I can search both files in one for loop, instead of two for loops (I tried to optimize the problem as much as possible). \n",
    "\n",
    "Assuming both files have the same length, I make a for loop through the index (not the values) so I can access both lists in one loop. I then check if the first column of both files are the same, in the same line. If the values are the same, the loop continues to the next row.If they are not the same, the program prints a statement telling that they are not the same results and returns false.\n",
    "\n",
    "If the program has not returned false, that means that all rows are equal and finally the program prints the statement that the rows are in the same sequence and returns True."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus tags in Germplasm.tsv and LocusGene.tsv are in the same sequence\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "def control_sequence(data_file1, data_file2, sep = '\\t'):\n",
    "    '''\n",
    "    This function compares the first column of two data files and\n",
    "    verifies that the sequences are in the same order.\n",
    "    The output is a true or false statement\n",
    "    '''\n",
    "    \n",
    "    # Read data from file 1 and file 2\n",
    "    with open(data_file1) as file1, open(data_file2) as file2:\n",
    "        \n",
    "        # Open the data from file1 and file two and convert it into a list\n",
    "        data1 = list(csv.reader(file1, delimiter=sep))\n",
    "        data2 = list(csv.reader(file2, delimiter=sep))\n",
    "        \n",
    "        # for loop by index in both files\n",
    "        for index in range(len(data1)):\n",
    "            \n",
    "            # Check if the locus tag of both files in the same row are the same,\n",
    "            if data1[index][0] == data2[index][0]:\n",
    "                # If true, go to next row\n",
    "                continue\n",
    "            # Case where the locus tag is not the same in both files\n",
    "            else:\n",
    "                print('Locus tags in ' + data_file1 + ' and ' + data_file2 + ' are not in the same sequence')\n",
    "                return False\n",
    "        \n",
    "        # If all files have the same sequence of locus tag, the program will reach this point\n",
    "        # and it means that both files have the same sequence\n",
    "        print('Locus tags in ' + data_file1 + ' and ' + data_file2 + ' are in the same sequence')\n",
    "        return True\n",
    "        \n",
    "# ------- MAIN CODE ----------   \n",
    "\n",
    "control_sequence('Germplasm.tsv', 'LocusGene.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To solve this problem I will use sqlMagic. I first connect to my local mysql. Then I create a database that I called exam2. Then I use that database and start creating the tables.\n",
    "\n",
    "The first table is germplasm and will have the information of Germplasm.tsv. This table will have four columns, the first one is the Locus Tag that will be the primary key because it is a unique identifier of each row, and we proved that both tables have the same unique locus tags. This locus tag is the same in the other table and is the column that will connect both tables in a 1:1 relationship, because each row in the germplasm table will have one corresponding row in the locus_gene table. The second column is a VARCHAR of the germplasm, the third column is also a VARCHAR(500) of the phenotype, which has more length. The fourth column is an INTEGER of the pubmed_id. \n",
    "\n",
    "The second table is locus_gene and has the information of LocusGene.tsv. The first column is the locus tag that is unique and it is the primary key of this table and will be used to link it to the germplasm table. The second column is a VARCHAR of the gene. The third column is a INTEGER of the protein length.\n",
    "\n",
    "It is important to note that this database design works because we are assuming that the locus tag is unique in each table. If this is not the case, I would use an id column with autoincrement set to true to identify each row.\n",
    "\n",
    "Also, all the columns in both tables are NOT NULL, forcing the database to be complete."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database exam2;\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use exam2;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_exam2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus_gene</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('germplasm',), ('locus_gene',)]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(10)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(50)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(500)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed_id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(10)', 'NO', 'PRI', None, ''),\n",
       " ('germplasm', 'varchar(50)', 'NO', '', None, ''),\n",
       " ('phenotype', 'varchar(500)', 'NO', '', None, ''),\n",
       " ('pubmed_id', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql DROP table germplasm\n",
    "%sql CREATE TABLE germplasm(locus VARCHAR(10) PRIMARY KEY NOT NULL, germplasm VARCHAR(50) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed_id INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql DROP table locus_gene\n",
    "%sql CREATE TABLE locus_gene(locus VARCHAR(10) PRIMARY KEY NOT NULL, gene VARCHAR(20) NOT NULL, protein_length INTEGER NOT NULL);\n",
    "%sql DESCRIBE locus_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To fill the database exam2 I will use pymysql, so I use the library pymysql to import the cursors method. I first stablish the connection to the exam2 database as a localhost, using my root credentials. \n",
    "\n",
    "Since I will need to read both files to fill the database, I will create an auxiliary function called dict_data_converter that uses the csv library and reads the file with a separator and converts it into a list of dictionaries using the method csv.DictReader. The arguments to this function is the name of the file and the separator, that by default is a tab delimited file.\n",
    "\n",
    "I try to connect to the exam2 database and read the data from the Germplasm.tsv file. I use a for loop to go through each line of the file. Each row is a dictionary with the key as the column header and the value is the row value. I then insert each row into the database using the SQL command INSERT INTO and using the appropiate key of the dictionary.\n",
    "\n",
    "I do the same procedure with the LocusGene.tsv file to fill in the locus_gene table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "def dict_data_converter(file, sep = '\\t'):\n",
    "    '''This function reads file and converts it into a list of dictionaries'''\n",
    "    # Open file\n",
    "    with open(file) as data:\n",
    "        # Convert to list of dictionaries\n",
    "        data_list = list(csv.DictReader(data, delimiter=sep))\n",
    "        return data_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "()\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        # Read the Germplasm.tsv file\n",
    "        germplasm_data = dict_data_converter('Germplasm.tsv')\n",
    "\n",
    "        # Go through every row in the germplasm table\n",
    "        for germplasm_row in germplasm_data:\n",
    "            # Create SQL statement\n",
    "            sql = \"INSERT INTO germplasm (locus, germplasm, phenotype, pubmed_id) VALUES ('{}','{}','{}',{})\".format(germplasm_row['Locus'], germplasm_row['germplasm'],germplasm_row['phenotype'], germplasm_row['pubmed'])\n",
    "            cursor.execute(sql) # Execute SQL\n",
    "        \n",
    "        # Read the LocusGene.tsv file\n",
    "        locus_data = dict_data_converter('LocusGene.tsv')\n",
    "        \n",
    "        # Go through every row in the locus_gene table\n",
    "        for locus_data_row in locus_data:\n",
    "            # Create SQL statement\n",
    "            sql = \"INSERT INTO locus_gene (locus, gene, protein_length) VALUES ('{}','{}',{})\".format(locus_data_row['Locus'], locus_data_row['Gene'], locus_data_row['ProteinLength'])\n",
    "            cursor.execute(sql) # Run SQL\n",
    "        \n",
    "        # To make sure program runs completely\n",
    "        print(\"The database loaded correctly\")\n",
    "        \n",
    "finally:\n",
    "    print(\"There was a problem loading the values to the database\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678)]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the germplasm table was filled correctly\n",
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the locus_gene table was filled correctly\n",
    "%sql SELECT * FROM locus_gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will solve this problem using pymysql. I stablish a connection with the exam2 database and try to connect with the cursor.\n",
    "\n",
    "I created a file called report.txt that will have the four reports generated in the code. I started the file with a header to describe the content of the report.\n",
    "\n",
    "To solve the problem 4.1 I made an SQL statement that selects all the data from both tables where (Inner join) the locus are the same, since that is the primary key it will join both tables into one with all the information. Once I have that query I execute it and fetch the result as a list of dictionaries, where the keys are the column names. That way I use a for loop to go through each row of the table and create a body variable for the report that has the text that describes each row.\n",
    "\n",
    "The problem 4.2 is solved in a similar way. The SQL statement is an explicit INNER JOIN (I could also use where) that joins both tables on the locus (primary key), and also the gene from the locus_gene table must be either SKOR or MAA3. Then I executed and wrote a report in a similar way than in problem 4.1.\n",
    "\n",
    "To solve problem 4.3 I generated a list of the number in string of each chromosome. I then made a for loop that goes through each chromosome and counts (COUNT()) all the entries in table germplasm (I could also use locus_gene since it has the same locus tags) that have the regex expression '^AT{chromosome number}G' using the command REGEXP. I anchor it to the beggining of the line and I know that the locus tags has the pattern ATXG, where X is the chromosome number. The count table result has just one entry, so I access it with[0] and use the name given by AS. I then run the query, fetch the results, and write the report.\n",
    "\n",
    "The problem 4.4 was solved in a similar way, but this time I used the statement AVG(locus_gene.protein_length) in the locus_gene table and made a for loop through each chromosome to get the average of the protein length in each chromosome (using the same REGEXP expression as problem 4.3). I then made the query, fetched the results and wrote the report.\n",
    "\n",
    "NOTE: Bellow the code there is the SQL magic of the statements that generates the tables for each problem. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "problem 4.1 OK\n",
      "Problem 4.2 OK\n",
      "Problem 4.3 OK\n",
      "Problem 4.4 OK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "#connection.autocommit = False  # note that it is possible to delay putting changes into the database!\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        # Open the report.txt file\n",
    "        report = open('report.txt', 'w')\n",
    "        \n",
    "        # Write a header line\n",
    "        header = \"Report of the germplasm and locus gene information - Generated by Guillermo Chumaceiro\\n\\n\"\n",
    "        report.write(header)\n",
    "        \n",
    "        # Problem 4.1\n",
    "        sql = \"\"\"SELECT * FROM germplasm, locus_gene WHERE \n",
    "        germplasm.locus = locus_gene.locus;\"\"\"\n",
    "        \n",
    "        cursor.execute(sql) #Execute SQL\n",
    "        results = cursor.fetchall() #  is list of matching results,  inside there is a dictionary. Keys are column names\n",
    "        \n",
    "        report.write('Answer to problem 4.1\\n')\n",
    "        # Go throught every row of the result query\n",
    "        for row in results:\n",
    "            body = 'The Locus Tag {} has the following information:\\n'.format(row['locus'])\n",
    "            body += '- The gene name: {}\\n'.format(row['gene'])\n",
    "            body += '- The germplasm: {}\\n'.format(row['germplasm'])\n",
    "            body += '- The description of the phenotype: {}\\n'.format(row['phenotype'])\n",
    "            body += '- The length of the protein encoded by {} is {} aminoacids long\\n'.format(row['gene'], row['protein_length'])\n",
    "            body += '- The pubmed ID: {}\\n'.format(row['pubmed_id'])\n",
    "            body += '\\n'\n",
    "            report.write(body)\n",
    "        print(\"problem 4.1 OK\")\n",
    "        \n",
    "        # Problem 4.2\n",
    "        sql = \"\"\"SELECT * FROM germplasm INNER JOIN locus_gene \n",
    "        ON germplasm.locus = locus_gene.locus \n",
    "        AND (locus_gene.gene = 'SKOR' OR locus_gene.gene = 'MAA3');\"\"\"\n",
    "        \n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        \n",
    "        report.write('Answer to problem 4.2\\n')\n",
    "        # Go throught every row of the result query\n",
    "        for row in results:\n",
    "            body = 'The gene {} has a Locus tag of {} and the germplasm {}. The protein it encodes has a length of {} aminoacids and it\\'s pubmed ID is {}. The phenotype described for this gene is: {}\\n'.format(row['gene'], row['locus'], row['germplasm'], row['protein_length'], row['pubmed_id'], row['phenotype'])\n",
    "            report.write(body)\n",
    "        print(\"Problem 4.2 OK\")\n",
    "        \n",
    "        # Problem 4.3\n",
    "        \n",
    "        report.write('\\nAnswer to problem 4.3\\n')\n",
    "        \n",
    "        # Generate a list with the five chromosome characters (can be any number, change range)\n",
    "        chromosomes = [str(i) for i in range(1,6)]\n",
    "        \n",
    "        # Go through the chromosomes list\n",
    "        for chromosome in chromosomes:\n",
    "            sql = \"\"\"SELECT COUNT(*) AS 'Chromosome_{0}' FROM germplasm WHERE locus REGEXP '^AT{0}G'\"\"\".format(chromosome)\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            \n",
    "            body = 'There are {} entries in Chromosome {}\\n'.format(results[0]['Chromosome_'+chromosome],chromosome)\n",
    "            report.write(body)\n",
    "        print(\"Problem 4.3 OK\")\n",
    "            \n",
    "        # Problem 4.4\n",
    "        report.write('\\nAnswer to problem 4.4\\n')\n",
    "        \n",
    "        # Loop though each chromosome\n",
    "        for chromosome in chromosomes:\n",
    "            sql = \"\"\"SELECT AVG(locus_gene.protein_length) AS 'Chrormosome_{0}_protein_average_length' FROM locus_gene WHERE locus REGEXP '^AT{0}G'\"\"\".format(chromosome)\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            \n",
    "            body = 'For the genes in Chromosome {} the average protein lenght is {} aminoacids\\n'.format(chromosome, results[0]['Chrormosome_'+ chromosome +'_protein_average_length'])\n",
    "            report.write(body)\n",
    "        print(\"Problem 4.4 OK\")\n",
    "        # close the report.txt file\n",
    "        report.close()\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed_id</th>\n",
       "        <th>locus_1</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351, 'AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401, 'AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125, 'AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744, 'AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956, 'AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505, 'AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900, 'AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276, 'AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837, 'AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180, 'AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444, 'AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653, 'AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278, 'AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843, 'AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770, 'AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667, 'AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018, 'AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690, 'AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066, 'AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425, 'AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186, 'AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662, 'AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499, 'AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832, 'AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666, 'AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749, 'AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244, 'AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216, 'AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401, 'AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251, 'AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687, 'AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678, 'AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SQL table result of problem 4.1\n",
    "%sql SELECT * FROM germplasm, locus_gene WHERE germplasm.locus = locus_gene.locus;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed_id</th>\n",
       "        <th>locus_1</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770, 'AT3G02850', 'SKOR', 234),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186, 'AT4G15570', 'MAA3', 294)]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SQL table result from problem 4.2\n",
    "%sql SELECT * FROM germplasm INNER JOIN locus_gene ON germplasm.locus = locus_gene.locus AND (locus_gene.gene = 'SKOR' OR locus_gene.gene = 'MAA3');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Chr1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(32,)]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SQL for problem 4.3\n",
    "%sql SELECT COUNT(*) AS 'Chr1' FROM germplasm WHERE locus REGEXP '^AT[1-5]G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Chrormosome_X_protein_average_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>258.8750</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(Decimal('258.8750'),)]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SQL for problem 4.4\n",
    "%sql SELECT AVG(locus_gene.protein_length) AS 'Chrormosome_X_protein_average_length' FROM locus_gene WHERE locus REGEXP '^AT[1-5]G'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
